seed/models/building_file.py

```diff
@@ -329,18 +331,34 @@ def process(self, organization_id, cycle, property_view=None):
                     meter_type = None
                 meter_conversions = self._kbtu_thermal_conversion_factors().get(meter_type, {})
 
-                valid_reading_models = {
-                    MeterReading(
+                for mr in valid_readings:
+                    # NOTE: This is really slow and a hack! Refactor this!
+                    # How we (as of writing) batch insert readings elsewhere: https://github.com/SEED-platform/seed/blob/6a31f3d369af1348afccb14c33ff9b4925cb711b/seed/data_importer/tasks.py#L849
+                    # More info: basically we can't update existing meter readings b/c
+                    # of our current unique_together. Whenever we try to update an instance
+                    # Django creates a WHERE clause that only selects based on the primary key (which is start_time)
+                    # which results in IntegrityErrors. Using a QuerySet's update method gets around this issue
+                    # if we filter on all of the fields we have a unique constraint on...
+                    defaults = {
+                        'reading': float(mr.get('reading', 0)) * meter_conversions.get(mr.get('source_unit'), 1.00),
+                        'source_unit': mr.get('source_unit'),
+                        'conversion_factor': meter_conversions.get(mr.get('source_unit'), 1.00),
+                    }
+                    num_updated = MeterReading.objects.filter(
+                        meter_id=meter.id,
                         start_time=mr.get('start_time'),
                         end_time=mr.get('end_time'),
-                        reading=float(mr.get('reading', 0)) * meter_conversions.get(mr.get('source_unit'), 1.00),
-                        source_unit=mr.get('source_unit'),
-                        meter_id=meter.id,
-                        conversion_factor=meter_conversions.get(mr.get('source_unit'), 1.00)
-                    )
-                    for mr in valid_readings
-                }
-                MeterReading.objects.bulk_create(valid_reading_models)
+                    ).update(**defaults)
+                    assert num_updated == 0 or num_updated == 1
+                    if num_updated == 0:
+                        # create the reading
+                        new_values = {
+                            'meter_id': meter.id,
+                            'start_time': mr.get('start_time'),
+                            'end_time': mr.get('end_time'),
+                        }
+                        new_values.update(defaults)
+                        MeterReading.objects.create(**new_values)
```

Comment:
I started typing one thing but then decided it might not be right (see below). TLDR I think it's worth reverting this to the original as you suggested, and writing a bunch of tests to verify meters are handled how we expect, then fix as needed.


ORIGINAL THOUGHTS:
Oof great question... Meters are a actually a little different from Scenarios and Measures -- instead of getting linked to the State or View, they are directly linked to the Canonical Property... Which means we might not want to just drop the old meters and readings...
@nllong what do you think? Here are the cases I can imagine:
1. User has existing meter with readings at times A, B, C. Then imports bsync which matches that meter id, and only has a new, non-overlapping reading D. I'd think we would add this reading to the existing meter
2. User has existing meter with readings at times A, B, C. Then imports bsync which matches that meter id, and has a new reading which starts before B and ends after B (i.e. consumes B's span). I'm not sure what we do here, do we enforce/check overlapping readings anywhere/should we be? If we don't care, we just create a new reading.
3. Same as 2, but instead of full overlap, it is a partial overlap with an existing reading.